### ISSUE TITLE
Bazel 0.1.2 does not build tensorflow

### ISSUE THREAD

**Issue by jendap** [bazelbuild/bazel#714]
bazel 0.1.2 deb package fails to build tensorflow

See commit https://github.com/jendap/tensorflow/commit/28996fa9f11dfce41a796d8055269f972e13acca

It gives errors like [this](http://ci.tensorflow.org/job/experimental-jendap-tensorflow/5/console)...
ERROR: /root/.cache/bazel/_bazel_root/68a62076e91007a7908bc42a32e4cff9/external/re2/BUILD:9:1: undeclared inclusion(s) in rule '@re2//:re2':
this rule is missing dependency declarations for the following files included by 'external/re2/re2/perl_groups.cc':
...

... or ...

ERROR: /tensorflow/google/protobuf/BUILD:29:1: undeclared inclusion(s) in rule '//google/protobuf:protobuf_lite':
this rule is missing dependency declarations for the following files included by 'google/protobuf/src/google/protobuf/stubs/once.cc':
...

Steps to reproduce (it is slow and needs good internet, it will build 5GB docker image):
```bash
git clone --recursive https://github.com/jendap/tensorflow.git
cd tensorflow
git checkout ci_build-bazel_upgrade
tensorflow/tools/ci_build/ci_build.sh ANDROID tensorflow/tools/ci_build/builds/android.sh
```

**Comment by jendap** [comment]
Note: this is different than those bugs about crosstool_wrapper_driver_is_not_gcc.

This build already runs with `--spawn_strategy=standalone`. So I guess it is different issue.

**Comment by damienmg** [comment]
It should be fixed in latest tensorflow repo, do you have a full output?

**Comment by jendap** [comment]
yes, it is on top of current master https://github.com/tensorflow/tensorflow/commit/6936918ad8abd7ade445674d7c912e69157a1017
https://github.com/jendap/tensorflow/commits/ci_build-bazel_upgrade
https://github.com/tensorflow/tensorflow/commits/master

Full output is http://ci.tensorflow.org/job/experimental-jendap-tensorflow/5/console

**Comment by damienmg** [comment]
Gosh. This is embarrassing: Can you try wiping Bazel's cache? There was a bug in Bazel that do not correctly invalidate the WORKSPACE file (I think it got fixed for all repository with https://github.com/bazelbuild/bazel/commit/f1ffe3621002231fe4828fe4ecfd5f64df9d2052).

I am still building the docker image to confirm.

Note that 0.1.3rc1 includes that fix, maybe you want to give it a try? :) What could be done to validate Bazel release candidate with all the buiilds on TensorFlow ci? We will never be able to have all the configuration of TensorFlow on ci.bazel.io.

**Comment by damienmg** [comment]
Arf I am on my mac with boot2docker and of course it cannot mount the volume. I will retry tomorrow if clearing the cache did not solved the issue

**Comment by jendap** [comment]
Good point, I should write a few lines to the readme how to run it with boot2docker. Later.

I'm going to rerun it with cache and with cleared cache...

I can create validation job for all the important tensorflow targets with a given bazel repo commit sha (or brach, tag) on linux easily. Other platform may be a bit trickier.

**Comment by jendap** [comment]
Clearing the cache did not helped 0.1.2, see...

It run on https://github.com/jendap/tensorflow/commits/ci_build-bazel_upgrade (latest master from official repo + the one small commit changing bazel) 

First build (cache about 1.5GB from previous builds) http://ci.tensorflow.org/job/experimental-jendap-tensorflow/8/console

Second build (after `rm -rf ~/.cache/bazel`) http://ci.tensorflow.org/job/experimental-jendap-tensorflow/9/console

Note: It runs fast as the docker containers (both with 0.1.1 and 0.1.2) are cached already.

**Comment by damienmg** [comment]
Oh having more slept helped more

@lberki this is related to the C++ Android Crosstool. It seems like the includes are caught by the strict header deps checker

**Comment by jendap** [comment]
The release-0.1.3 branch fails the same way. http://ci.tensorflow.org/job/experimental-jendap-tensorflow/45/console

**Comment by damienmg** [comment]
Ok reproduced inside the docker container but cannot reproduce it outside. We will spend some more time tomorrow with Lukacs to find out what exactly is going on.

**Comment by jendap** [comment]
Interesting! The environment inside docker should be even better defined. The biggest difference may be the [.bazelrc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/install/.bazelrc).

I can try it tomorrow myself.

**Comment by damienmg** [comment]
The .bazelrc shouldn't be a problem. I am guessing that it triggers some
bug related to the path of the workspace or the fact that bazel runs as
root. I will do some more test when I arrive at work.

**Comment by jendap** [comment]
Master no longer runs as root inside the docker due to https://github.com/tensorflow/tensorflow/commit/def98063aa75e76f042558d0dbc65e6236d97ad7. It now pick user group and home directory from user calling ci_build.sh (jenkins on ci.tensorflow.org). I have to yet test it with upgraded bazel.

**Comment by damienmg** [comment]
Ok I have tried many different things and cannot reproduce that outside of docker. I am leaning toward the version of the android sdk now.

**Comment by damienmg** [comment]
We did some debugging with @lberki  and the issues seems that Bazel writes wrong CROSSTOOL protos for the Android NDK. I particular, it says at line ~1086 in `external/androidndk/BUILD`:

```
cxx_builtin_include_directory: "external/androidndk/ndk/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9/include"
  cxx_builtin_include_directory: "external/androidndk/ndk/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9/include-fixed"
  cxx_builtin_include_directory: "%sysroot%/usr/include"
unfiltered_cxx_flag: "-isystem"
  unfiltered_cxx_flag: "external/androidndk/ndk/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9/include"
  unfiltered_cxx_flag: "-isystem"
  unfiltered_cxx_flag: "external/androidndk/ndk/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9/include-fixed"
  unfiltered_cxx_flag: "-isystem"
  unfiltered_cxx_flag: "external/androidndk/ndk/sources/cxx-stl/gnu-libstdc++/4.9/include"
  unfiltered_cxx_flag: "-isystem"
  unfiltered_cxx_flag: "external/androidndk/ndk/sources/cxx-stl/gnu-libstdc++/4.9/libs/armeabi-v7a/include"
  unfiltered_cxx_flag: "-isystem"
  unfiltered_cxx_flag: "external/androidndk/ndk/sources/cxx-stl/gnu-libstdc++/4.9/include/backward"
```

Which means that a few directories in `-isystem` are not ignored for header dependency checks, including the ones Bazel complains about:

```
'/opt/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/include/string'
'/opt/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/libs/armeabi-v7a/include/bits/c++config.h'
'/opt/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/include/bits/stringfwd.h'
'/opt/android-ndk-r10e/platforms/android-21/arch-arm/usr/include/wchar.h'
```

There is also the issue that gcc might do an extra level of symlink resolution which we don't want it to. 

This is fixable but will require some effort from @lberki , mainly to reproduce this preferably outside of a Docker container so that he doesn't go bonkers debugging this (it's a mystery why this issue doesn't arise outside a Docker container)

**Comment by lberki** [comment]
Assigning this to Alex temporarily since I'm not sure how much network access I'll get tomorrow given that the office will be closed. If the move in NYC goes more smoothly, could you take a look?

**Comment by lberki** [comment]
Hm, this is strange. I managed to reproduce this on my home laptop within the Docker container, then copied both the Bazel workspace and SDK and downloaded the same NDK out of the container so that I may debug it easier, and then it suddenly started working. The same NDK/SDK and the same workspace. This will require some more investigation.

My guess is that gcc sometimes decides to resolve symlinks for the .d files and sometimes it doesn't and that confuses Bazel.

**Comment by lberki** [comment]
Oh, I see. The sandbox is not active in Docker and it does fail without a container when using `--spawn_strategy=standalone`.

**Comment by lberki** [comment]
I have a patch! It probably breaks all the tests because my work environment today is quite hostile, but it does work:

https://gist.github.com/lberki/d49d42f7abe9369ef305

**Comment by jendap** [comment]
I'm not sure what was your base commit. I can't apply it cleanly to test it. It's better share a branch even in your personal repo than patch in gist ;-)

**Comment by lberki** [comment]
HEAD on Bazel master. I admit to being somewhat sloppy here. In my defense, I was quite happy that I could even get that far. Let's see if I can finish this up today.

**Comment by lberki** [comment]
I made a bit more thorough investigation of the behavior of gcc and it's odd (to say the least). Having the Android NDK in these directories does not work:

```
/opt
/a
/ab
/abc
/abcd
/a/b
```

However, having it in these directories *does* work:

```
/opt/a
/opt/a/b
/a/b/c
/abcde
```

So I conclude that the gcc in the NDK behaves differently if the directory where the NDK is put has 5 characters or less (including the beginning slash). I'll send out a change, but maybe it'd be better to tweak the build process of Tensorflow instead of putting this hack in Bazel?

**Comment by jendap** [comment]
Number of characters in path makes a difference? That is weird! 

Is tensorflow doing something wrong in BUILD files? If it has to be a hack in one or another project I would suggest to put it in bazel as other projects may run into the same issue.

@lberki please let us know how to fix it if you think it is tensorflow only issue. @vrv has also asked for a fix.

**Comment by lberki** [comment]
@jendap : it should be enough to put the Android NDK in your docker container under a longer path name than `/opt/android-ndk-r10e`. E.g. `/opt/dependencies/android-ndk-r10e` would work well (as far as I can tell. It's weird.). I'd give it a stab, except that the Munich office was moving, it's Christmas holiday and I don't know much about the setup of your Docker container.

**Comment by jendap** [comment]
@lberki you are right about the five characters. I have just tested it. That is weird. What are you doing in bazel? :-) I can put the workaround in the tensorflow ci. But obviously tensorflow (and other projects) users will run to the such a problem. I personally have android sdk and ndk in /opt.

BTW: I'm sorry for the noise from my test comments. I have not realized github is automatically linking them.

**Comment by lberki** [comment]
@jendap this is not specific to Bazel, it's a weirdness of gcc: it resolves symlinks in the .d files iff the path to the Android NDK is 5 characters or less. Why, I have no clue.

**Comment by jendap** [comment]
@lberki Not Bazel? Why does older bazel work with the same version of android ndk? Docker container with every single file on the disk the same except /bazel works for 0.1.1 but fails for 0.1.2.

The workaround https://github.com/tensorflow/tensorflow/commit/2e680bbcace962a7aa85baaef948d360e6a2c49b (/android instead of /opt) has landed in tensorflow continuous integration. We can use there newer bazel.

BTW:  I have played a bit with a bisect job. I will try to find some time to finish it next weekend. The idea is to have one jenkins job able to run git bisect on either tensorflow or jenkins repo. That should help with such issues.

**Comment by skgumma** [comment]
I'm having the same issue. My environment:
Cent OS 6.6
bazel version
Build label: 0.3.0- (@non-git)
Build target: bazel-out/local-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Tue Jul 26 02:34:53 2016 (1469500493)
Build timestamp: 1469500493
Build timestamp as int: 1469500493
GCC: 4.9.3 installaed in custom location. 

Compilation error message:

ERROR: /root/.cache/bazel/_bazel_root/7f66018f0a2ea01f86fb61068feb05d8/external/protobuf/BUILD:71:1: undeclared inclusion(s) in rule '@protobuf//:protobuf_lite':
this rule is missing dependency declarations for the following files included by 'external/protobuf/src/google/protobuf/stubs/atomicops_internals_x86_msvc.cc':
  '/app/gcc/4.9.3/lib/gcc/x86_64-unknown-linux-gnu/4.9.3/include/stdarg.h'
  '/app/gcc/4.9.3/lib/gcc/x86_64-unknown-linux-gnu/4.9.3/include/stddef.h'
  '/app/gcc/4.9.3/lib/gcc/x86_64-unknown-linux-gnu/4.9.3/include/stdint.h'
  '/app/gcc/4.9.3/lib/gcc/x86_64-unknown-linux-gnu/4.9.3/include-fixed/limits.h'
  '/app/gcc/4.9.3/lib/gcc/x86_64-unknown-linux-gnu/4.9.3/include-fixed/syslimits.h'.
Target //tensorflow/tools/pip_package:build_pip_package failed to build

**Comment by damienmg** [comment]
which version of Bazel exactly? It somewhere after 0.3.0 but there is not a git hash to identify.

I would say it because you need to modify the third_party/gpus/crosstool/CROSSTOOL file to add /app/gcc/4.9.3/lib/gcc/x86_64-unknown-linux-gnu/4.9.3/include in the list of cxx_include_directories

**Comment by skgumma** [comment]
Thanks @damienmg 
I have these lines in my CROSSTOOL file  and still I see the compilation failing. 

# grep cxx tensorflow/third_party/gpus/crosstool/CROSSTOOL
  cxx_flag: "-std=c++11"
  cxx_builtin_include_directory: "/app/gcc/4.9.3"
  cxx_builtin_include_directory: "/app/gcc/4.9.3/include"
  cxx_builtin_include_directory: "/app/gcc/4.9.3/lib/gcc/x86_64-unknown-linux-gnu/4.9.3/include"
  cxx_builtin_include_directory: "/app/cuda/7.5/include"
  unfiltered_cxx_flag: "-no-canonical-prefixes"
  unfiltered_cxx_flag: "-Wno-builtin-macro-redefined"
  unfiltered_cxx_flag: "-D__DATE__=\"redacted\""
  unfiltered_cxx_flag: "-D__TIMESTAMP__=\"redacted\""
  unfiltered_cxx_flag: "-D__TIME__=\"redacted\""
  unfiltered_cxx_flag: "-fno-canonical-system-headers"
  cxx_flag: "-std=c++11"
  cxx_builtin_include_directory: "/"
  unfiltered_cxx_flag: "-no-canonical-prefixes"
  unfiltered_cxx_flag: "-Wno-builtin-macro-redefined"
  unfiltered_cxx_flag: "-D__DATE__=\"redacted\""
  unfiltered_cxx_flag: "-D__TIMESTAMP__=\"redacted\""
  unfiltered_cxx_flag: "-D__TIME__=\"redacted\""

**Comment by damienmg** [comment]
What does `realpath /app/gcc/4.9.3/lib/gcc/x86_64-unknown-linux-gnu/4.9.3/include/stdarg.h` says?

**Comment by skgumma** [comment]
I see the file there

# ls /app/gcc/4.9.3/lib/gcc/x86_64-unknown-linux-gnu/4.9.3/include/stda*
/app/gcc/4.9.3/lib/gcc/x86_64-unknown-linux-gnu/4.9.3/include/stdalign.h
/app/gcc/4.9.3/lib/gcc/x86_64-unknown-linux-gnu/4.9.3/include/stdarg.h
/app/gcc/4.9.3/lib/gcc/x86_64-unknown-linux-gnu/4.9.3/include/stdatomic.h

**Comment by damienmg** [comment]
I wanted to know if they are symlinks on the path, that might be the issue

**Comment by skgumma** [comment]
No there's no symlinks, it's my direct path.

**Comment by damienmg** [comment]
Can you provide the whole CROSSTOOL file?

Can you also confirm you are building with --config cuda?

**Comment by skgumma** [comment]
I'm using the below command to compile:

bazel build -c opt --config=cuda //tensorflow/tools/pip_package:build_pip_package --verbose_failures

Attaching the CROSSTOOL file. (Renamed as .txt just to ensure I can upload the file)
[CROSSTOOL.txt](https://github.com/bazelbuild/bazel/files/391745/CROSSTOOL.txt)

**Comment by skgumma** [comment]
@damienmg Appreciate if you can help me install. I got struck here and it's very important for me to install tensorflow. Thanks.

**Comment by damienmg** [comment]
@skgumma: Are you on OS X?

Can you try replacing your cxx_* lines with only:
```
  cxx_builtin_include_directory: "/app”
```

**Comment by skgumma** [comment]
@damienmg I'm using CentOS 6.6. Well I got some help from my colleague who managed to compile bazel. The issue for us is each script is calling env python so every time my default environment is loaded instead of compile environment. 
He  replaced the default python with his own wrapper which does sets the environment variables and calls python which was installed in non system default location.

file name is: tensorflow/third_party/gpus/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc
First line in the script was #!/usr/bin/env python and replaced with  #!/usr/bin/env python-wrapper
python-wrapper is in tensorflow/ directory

Content of python-wrapper is:
module load python/2.7.11 gcc/4.9.3 binutils cuda/7.5

export JAVA_HOME=/opt/java/default

export EXTRA_BAZEL_ARGS="-s --verbose_failures --ignore_unsupported_sandboxing --genrule_strategy=standalone --spawn_strategy=standalone --jobs 1  --crosstool_top=//tools/cpp:toolchain --cpu=k8"
#EXTRA_BAZEL_ARGS="-s --verbose_failures --ignore_unsupported_sandboxing --genrule_strategy=standalone --spawn_strategy=standalone --jobs 8  --crosstool_top=//tools/cpp:toolchain --cpu=k8"

export PATH=/opt/bazel-0.3.0/output/:$PATH
export PATH=/opt/pcre/8.39/bin:$PATH
export LD_LIBRARY_PATH=/opt/pcre/8.39/lib:$LD_LIBRARY_PATH
export PATH=/opt/swig/gcc/3.0.10/bin:$PATH
export PATH=$PATH:/usr/bin

python "$@"

**Comment by damienmg** [comment]
/cc @aehlig this is another instance of env loading :)

TL;DR: We are fixing environment loading as a whole

**Comment by damienmg** [comment]
This should now be solved with --action_env
